---
layout: page
title: RNA-seq analysis
---

```{r options, echo=FALSE}
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

## Introduction

RNA-Seq is a valuable experiment for quantifying both the types and the amount of RNA molecules in a sample. We've covered the basic idea of the protocol in lectures, but some additional references for RNA-Seq include [Mortazavi (2008)](#foot) and [Marioni (2008)](#foot).

In the file, [cufflinks.txt](https://github.com/genomicsclass/labs/blob/master/course5/cufflinks.txt), we show some example calls for aligning RNA-Seq reads to the reference genome, and an example call for running Cufflinks, which estimates both the expression levels and the different RNA isoforms which are present in the sample. The main paper describing the Cufflinks method is [Trapnell (2010)](#foot) and the website for the Cufflinks software is:

<http://cufflinks.cbcb.umd.edu/>

In this lab, we will focus on comparing the expression levels of different samples, by counting the number of reads which overlap the exons of genes defined by a known annotation. As described in the lecture, this analysis sets aside the task of estimating the different kinds of RNA molecules.

## Counting reads in genes

We will work with a count matrix, which has genes along the rows and samples along the columns. The elements in the matrix give the number of reads which could be uniquely aligned to a given gene for a given sample. We will use count matrices already prepared, as otherwise, students would have to download very large BAM files containing the aligned reads. However, in [another lab](read_counting.html) we show the steps necessary to produce such a count matrix.

## Visualizing sample-sample distances

We will work with the [Hammer et al](#foot) dataset, as prepared by the ReCount website:

<http://bowtie-bio.sourceforge.net/recount/>

> ReCount is an online resource consisting of RNA-seq gene count datasets built 
> using the raw data from 18 different studies. The raw sequencing data (.fastq files) 
> were processed with Myrna to obtain tables of counts for each gene.

This is really helpful for us, so we don't have to download all the FASTQ files and map them ourselves. If you use this resource, you should cite [Frazee (2011)](#foot), and cite the appropriate paper for the experimental data that you download.

Here we read in the `Eset` hosted by ReCount, and turn it into a SummarizedExperiment.

```{r}
link <- "http://bowtie-bio.sourceforge.net/recount/ExpressionSets/hammer_eset.RData"
if (!file.exists("hammer_eset.RData")) download.file(link, "hammer_eset.RData")
load("hammer_eset.RData")
library(Biobase)
library(GenomicRanges)
# the SimpleList() part below is only necessary for Bioc <= 2.13
se <- SummarizedExperiment(SimpleList(counts=exprs(hammer.eset)))
colData(se) <- DataFrame(pData(hammer.eset))
```

We need to fix a typo in the Time column:

```{r}
colData(se)
colData(se)$Time[4] <- "2 months"
colData(se)$Time <- factor(colData(se)$Time)
colData(se)$Time
```

### Normalization

We will use the `DESeq2` package to normalize the sample for sequencing depth. For now, don't worry about the `design` argument.

```{r}
# biocLite("DESeq2")
library(DESeq2)
dds <- DESeqDataSet( se, design = ~ 1 )

```

The following estimates size factors to account for differences in sequencing depth.

```{r}
dds <- estimateSizeFactors( dds )
sizeFactors(dds)
colSums(counts(dds))
plot(sizeFactors(dds), colSums(counts(dds)))
abline(lm(colSums(counts(dds)) ~ sizeFactors(dds) + 0))
```

Now we can divide the columns by the size factor and take the log2 of these normalized counts plus a pseudocount of 1. We transpose in order to run PCA.

```{r}
logcounts <- log2( counts(dds, normalized=TRUE) + 1 )
pc <- prcomp( t( logcounts ) )
```

A couple EDA plots:

```{r}
library(rafalib)
mypar()
plot(pc$x[,1], pc$x[,2], 
     col=colData(dds)$protocol, 
     pch=as.numeric(colData(dds)$Time)+15)
plot(hclust(dist(t(logcounts))), labels=colData(dds)$protocol)
plot(hclust(dist(t(logcounts))), labels=colData(dds)$Time)
plot(logcounts[,1], logcounts[,2], cex=.1)
```

Now we will use a normalization method, which is similar to the variance stablizing normalization method mentioned in Week 5. It uses the variance model to shrink together the sample values for lowly expressed genes with high variance. 

The data is in the `assay` slot, and needs to be transposed as before to run PCA.

```{r}
# this takes ~15 seconds
rld <- rlog( dds )
# if using Bioc 2.13:
# rld <- rlogTransformation( dds )
pc2 <- prcomp( t( assay(rld) ) )
```

We can look at the same plots now using this transformed data.

```{r}
plot(pc2$x[,1], pc2$x[,2],
     col=colData(rld)$protocol, 
     pch=as.numeric(colData(rld)$Time)+15)
plot(hclust(dist(t(assay(rld)))), labels=colData(rld)$protocol)
plot(hclust(dist(t(assay(rld)))), labels=colData(rld)$Time)
plot(assay(rld)[,1], assay(rld)[,2], cex=.1)
```

## Differential gene expression

A number of methods for assessing differential gene expression from RNA-Seq counts use the Negative Binomial distribution to make probabilistic statements about the differences seen in an experiment. A few such methods are [edgeR](#foot), [DESeq](#foot), [DSS](#foot) and *many* others. A very incomplete list of other methods is provided in the [footnotes](#foot).

We will use `DESeq2` to perform differential gene expression on the counts. This also uses a Negative Binomial distribution to model the counts. It performs a similar step to `limma`, in using the variance of all the genes to improve the variance estimate for each individual gene. In addition, it shrinks the high variance fold changes, which will be seen in the resulting MA-plot.

First, we setup the `design` of the experiment, so that differences will be considered across time and protocol variables. The last variable is used for the default results tables and plots, and we make sure the "control" level is the first level, such that log fold changes will be treatment over control, and not control over treatment.

```{r}
colData(dds)$protocol
# if control was not already the "base level", we would do:
colData(dds)$protocol <- relevel(colData(dds)$protocol, "control")
levels(colData(dds)$protocol)
design(dds) <- ~ Time + protocol
```

The following line runs the model, and then we can extract a results table for all genes:

```{r}
# this takes ~20 seconds
dds <- DESeq( dds )
res <- results( dds )
head(res)
```

We can also make other results tables, such as control over SNL, or for comparing the time variable.

```{r}
head(results(dds, contrast=c("protocol","control","L5 SNL")))
head(results(dds, contrast=c("Time","2 months","2 weeks")))
```

We can now contruct an MA-plot of the fold change over the average expression level of all samples.

```{r}
# Bioc 2.13
plotMA(dds, ylim=c(-5,5))
# Bioc 2.14
plotMA(res, ylim=c(-5,5))
```

Suppose we are not interested in small log2 fold changes. We can also test for log2 fold changes larger than 1 in absolute value.

```{r}
resBigFC <- results(dds, lfcThreshold=1, altHypothesis="greaterAbs")
plotMA(resBigFC, ylim=c(-5,5))
abline(h=c(-1,1),lwd=5)
```

Let's examine the top gene, sorting by p-value:

```{r}
resSort <- res[order(res$pvalue),]
head(resSort)
k <- counts(dds)[rownames(resSort)[1],]
cond <- with(colData(se), factor(paste(Time, protocol)))
par(mar=c(15,5,2,2))
stripchart(log2(k + 1) ~ cond, method="jitter", vertical=TRUE, las=2)
```

We can then check the annotation of these highly significant genes:

```{r}
# biocLite("org.Rn.eg.db")
library(org.Rn.eg.db)
keytypes(org.Rn.eg.db)
head(rownames(dds))
geneinfo <- select(org.Rn.eg.db, keys=rownames(resSort)[1:20],
                   columns=c("ENSEMBL","SYMBOL","GENENAME"), 
                   keytype="ENSEMBL")
geneinfo
```

## Footnotes <a name="foot"></a>

### Introduction

Mortazavi A, Williams BA, McCue K, Schaeffer L, Wold B., "Mapping and quantifying mammalian transcriptomes by RNA-Seq", Nat Methods. 2008.
<http://www.nature.com/nmeth/journal/v5/n7/full/nmeth.1226.html>

John C. Marioni, Christopher E. Mason, Shrikant M. Mane, Matthew Stephens, and Yoav Gilad, "RNA-seq: An assessment of technical reproducibility and comparison with gene expression arrays" Genome Res. 2008.
<http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2527709/>

Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.,  "Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation", Nature Biotechnology, 2010.
<http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html>
<http://cufflinks.cbcb.umd.edu/>

### Hammer et al

Hammer P, Banck MS, Amberg R, Wang C, Petznick G, Luo S, Khrebtukova I, Schroth GP, Beyerlein P, Beutler AS. "mRNA-seq with agnostic splice site discovery for nervous system transcriptomics tested in chronic pain." Genome Res. 2010
<http://www.ncbi.nlm.nih.gov/pubmed?term=20452967>

### ReCount

Frazee AC, Langmead B, Leek JT. "ReCount: a multi-experiment resource of analysis-ready RNA-seq gene count datasets". BMC Bioinformatics 12:449 <http://www.ncbi.nlm.nih.gov/pubmed/22087737>

### Negative Binomial methods for differential expression of count data

All the following methods are available on Bioconductor:

- `edgeR`

Mark D. Robinson, Davis J. McCarthy, and Gordon K. Smyth, "edgeR: a Bioconductor package for differential expression analysis of digital gene expression data" Bioinformatics 2010.
<http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/>

- `DESeq` (the latest version is a separate package, `DESeq2`)

Simon Anders and Wolfgang Huber, "Differential expression analysis for sequence count data", Genome Biology 2010.
<http://genomebiology.com/2010/11/10/r106>

- `DSS`

Hao Wu, Chi Wang, Zhijin Wu, "A new shrinkage estimator for dispersion improves differential expression detection in RNA-seq data" Biostatistics 2013.
<http://biostatistics.oxfordjournals.org/content/14/2/232>

### Transformation followed by linear model methods

`voom` in the `limma` Bioconductor package

Charity W Law, Yunshun Chen, Wei Shi and Gordon K Smyth, "voom: precision weights unlock linear model analysis tools for RNA-seq read counts", Genome Biology. 2014.
<http://genomebiology.com/2014/15/2/R29>

### Resampling-based methods

`SAMseq` in the `samr` package on CRAN

Jun Li and Robert Tibshirani, "Finding consistent patterns: A nonparametric approach for identifying differential expression in RNA-Seq data", Stat Methods Med Res. 2013.
<http://smm.sagepub.com/content/22/5/519.short>

### Incorporating isoform-abundance

- `Cuffdiff` (the latest version is `Cuffdiff2`)

Trapnell C, Hendrickson DG, Sauvageau M, Goff L, Rinn JL, Pachter L., "Differential analysis of gene regulation at transcript resolution with RNA-seq" Nat Biotechnol. 2013.
<http://www.ncbi.nlm.nih.gov/pubmed/23222703>

- `BitSeq`

Peter Glaus, Antti Honkela, and Magnus Rattray, "Identifying differentially expressed transcripts from RNA-seq data with biological variation", Bioinformatics. 2012.
<http://bioinformatics.oxfordjournals.org/content/28/13/1721>

