---
layout: page
title: Basic Bioconductor infrastructure
---

```{r options, echo=FALSE}
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

## IRanges
Bring up the manual by calling
```{r}
browseVignettes("IRanges")
```
Irange has a start an end and a width. IRanges is the constructor for 
IRanges instances
```{r}
library(IRanges)
ir <- IRanges(5,10)
ir
start(ir)
end(ir)
width(ir)
# ?IRanges
```

```{r}
ir <- IRanges(start=c(3,5,17), end=c(10,8,20))
ir
ir <- IRanges(5,10)
```

```{r}
# ?"intra-range-methods"
shift(ir, -2)
```

Remeber, all of these commands can work on more than one range at once. Here we show the effects of the different methods using a single range:

```{r}
shift(ir,-2)
narrow(ir, start=2)
narrow(ir, end=5)
flank(ir, width=3, start=TRUE, both=FALSE)
flank(ir, width=3, start=FALSE, both=FALSE)
flank(ir, width=3, start=TRUE, both=TRUE)
ir * 2
ir + 2
ir - 2
```
+,- expands or contract both ends. * change the width (zoom in(positive) or zoom out(negative))

```{r}
# set up a plotting window so we can look at range operations
plotir <- function(ir,i) { arrows(start(ir)-.5,i,end(ir)+.5,i,code=3,angle=90,lwd=3) }
plot(0,0,xlim=c(0,15),ylim=c(0,11),type="n",xlab="",ylab="",xaxt="n")
axis(1,0:15)
abline(v=0:30 + .5,col=rgb(0,0,0,.5))

# plot the original IRange
plotir(ir,1)

# draw a red shadow for the original IRange
polygon(c(start(ir)-.5,start(ir)-.5,end(ir)+.5,end(ir)+.5),c(-1,12,12,-1),col=rgb(1,0,0,.2),border=NA)
plotir(shift(ir,-2), 2)
plotir(narrow(ir, start=2), 3)
plotir(narrow(ir, end=5), 4)
plotir(flank(ir, width=3, start=TRUE, both=FALSE), 5)
plotir(flank(ir, width=3, start=FALSE, both=FALSE), 6)
plotir(flank(ir, width=3, start=TRUE, both=TRUE), 7)
plotir(ir * 2, 8)
plotir(ir + 2, 9)
plotir(ir - 2, 10)
```

```{r}
# ?"inter-range-methods"
ir <- IRanges(start=c(3,5,17), end=c(10,8,20))
range(ir) #beginning and the end of the ranges including gaps. 
reduce(ir) # Does  not give the gaps
gaps(ir)
disjoin(ir)
```
Plot some diagrams
```{r}
plot(0,0,xlim=c(0,25),ylim=c(0,11),type="n",xlab="",ylab="",xaxt="n")
axis(1,0:25)
abline(v=0:30 + .5,col=rgb(0,0,0,.5))
plotir(ir,1)
plotir(range(ir),2)
plotir(reduce(ir),3)
plotir(gaps(ir),4)
plotir(disjoin(ir),5)
```

## GRanges and GRangesList

### GRanges

```{r}
library(GenomicRanges)
gr <- GRanges("chrZ", IRanges(start=c(5,10),end=c(35,45)),
              strand="+", seqlengths=c(chrZ=100L))
gr
shift(gr, 10)
shift(gr, 80) # goes off the boundary
trim(shift(gr, 80)) # trim the ranges
mcols(gr) #Metadata columns
mcols(gr)$value <- c(-1,4) #set values for metadata
gr
```

### GRangesList

```{r}
gr2 <- GRanges("chrZ",IRanges(11:13,51:53))
mcols(gr)$value <- NULL
grl <- GRangesList(gr,gr2)
grl
length(grl) #A list of GRanges. e.g. exons by genes or transcripts
grl[[1]]
mcols(grl)$value <- c(5,7)
grl
mcols(grl)#values are assigned to each granges object in the list
```

### findOverlaps and %over%

```{r}
gr1 <- GRanges("chrZ",IRanges(c(1,11,21,31,41),width=5))
gr2 <- GRanges("chrZ",IRanges(c(19,33),c(38,35)))
gr1
gr2
fo <- findOverlaps(gr1, gr2)
fo
queryHits(fo)
subjectHits(fo)
gr1 %over% gr2
gr1[gr1 %over% gr2]# Return those in gr1 that overlap with gr2
```

### Rle and Views

```{r}
r <- Rle(c(1,1,1,0,0,-2,-2,-2,rep(-1,20)))
r # store a compact reprentation of the sequence
str(r) 
as.numeric(r)
Views(r, start=c(4,2), end=c(7,6)) #Set windows on the sequence represented by the rle object r
```



## ExpressionSet and SummarizedExperiment

```{r}
library(Biobase)
library(GEOquery)
geoq <- getGEO("GSE9514")
names(geoq)
e <- geoq[[1]] #This has only one expression set element
```

### ExpressionSet
Generally used for arrays, rows are features
```{r}
e
dim(e)
exprs(e)[1:3,1:3] #retrieve microarray values of the first 3 genes for the first 3 samples
#first 5 genes for the first sample
head(exprs(e)[,1])
#first gene for all samples
exprs(e)[1,]
#indexing by name
exprs(e)["10001_at",]
#gene sets
rownames(exprs(e))
dim(exprs(e))
#phenotypes
pData(e)[1:3,1:6]
dim(pData(e))
names(pData(e))
pData(e)$characteristics_ch1 #strain information
#get the number of replicate for each strain
table(pData(e)$characteristics_ch1)
#make the names shorter
as.numeric(pData(e)$characteristics_ch1)
#features
fData(e)[1:3,1:3] #rows are genes, columns are annotation information
dim(fData(e))
names(fData(e))
head(fData(e)$"Gene Symbol")#4869 gene symbols for 9335 rows. 3722 have empty name "". Some are duplicated 
head(rownames(e)) #genes

experimentData(e)
annotation(e)
```

### Summarized Experiment
Generally used for sequencing, rows are GRanges
```{r}
library(parathyroidSE)
data(parathyroidGenesSE)
se <- parathyroidGenesSE
se
```


```{r}
dim(se)
assay(se)[1:3,1:3]
dim(assay(se))# this object contains only one assay

colData(se)[1:3,1:6]#equivalent to pData for microarray
dim(colData(se))
names(colData(se))
colData(se)$treatment
table(colData(se)$treatment)

rowData(se)[1]#genes as collections of exons represnted as GRanges
class(rowData(se))
length(rowData(se))#number of genes
head(rownames(se))
metadata(rowData(se))

exptData(se)$MIAME
abstract(exptData(se)$MIAME)
```

## Footnotes

For more information about the `GenomicRanges` package, check out the PLOS Comp Bio paper, which the authors of GenomicRanges published:

<http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118>

Also the software vignettes have a lot of details about the functionality. Check out "An Introduction to Genomic Ranges Classes":

<http://www.bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf>

All of the vignette PDFs are available here:

<http://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html>


