---
layout: page
title: NGS read counting
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

The following lab will describe how to count NGS reads which fall into genomic features. We want to end up with a count matrix which has rows corresponding to genomic ranges and columns which correspond to different experiments or samples. As an example, we will use an RNA-Seq experiment, with files in the `pasillaBamSubset` Bioconductor data package. However, the same functions can be used for DNA-Seq, ChIP-Seq, etc.

```{r, message=FALSE}
#biocLite("pasillaBamSubset")
#biocLite("TxDb.Dmelanogaster.UCSC.dm3.ensGene")
library(pasillaBamSubset)
library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
```

We load a transcript database object. These are prebuilt in R for various well studied organisms, for example `TxDb.Hsapiens.UCSC.hg19.knownGene`. In addition the `makeTranscriptDbFromGFF` file can be used to import GFF or GTF gene models. We use the `exonsBy` function to get a `GRangesList` object of the exons for each gene.

```{r}
txdb <- TxDb.Dmelanogaster.UCSC.dm3.ensGene
grl <- exonsBy(txdb, by="gene")
grl[100]
grl[[100]]
grl[[100]][1]
```

These functions in the `pasillaBamSubset` package just point us to the BAM files.

```{r}
fl1 <- untreated1_chr4()
fl2 <- untreated3_chr4()
fl1
```

We need the following libraries for counting BAM files.

```{r}
library(Rsamtools)
library(GenomicRanges)
library(GenomicAlignments)
```

We specify the files using the `BamFileList` function. For more fine grain control, you can tell the read counting functions how many reads to load at once with the `yieldSize` argument.

```{r}
fls <- BamFileList(c(fl1, fl2))
names(fls) <- c("first","second")
```

The following function counts the overlaps of the reads in the BAM files in the features, which are the genes of Drosophila. We tell the counting function to ignore the strand, i.e., to allow minus strand reads to count in plus strand genes, and vice versa.

```{r}
so1 <- summarizeOverlaps(features=grl,
                         reads=fls,
                         ignore.strand=TRUE)
so1
```

We can examine the count matrix, which is stored in the `assay` slot:

```{r}
head(assay(so1))
colSums(assay(so1))
```

The other parts of a `SummarizedExperiment`.

```{r}
rowData(so1) # rowData will become rowRanges in the next release of Bioc
colData(so1)
colData(so1)$sample <- c("one","two")
colData(so1)
metadata(rowData(so1)) 
```

We can do some basic exploratory data analysis of the counts:

```{r}
x <- assay(so1)[,1]
hist(x[x > 0], col="grey")
hist(x[x > 0 & x < 10000], col="grey")
plot(assay(so1) + 1, log="xy")
```

The second file should actually be counted in a special manner, as it contains pairs of reads which come from a single fragment. We do not want to count these twice, so we set `singleEnd = FALSE`. Additionally, we specify `fragments = TRUE` which counts reads if only one of the pair aligns to the features, and the other pair aligns to no feature.

```{r}
# ?untreated3_chr4
# ?summarizeOverlaps
fls <- BamFileList(fl2)
so2 <- summarizeOverlaps(features=grl,
                         reads=fls,
                         ignore.strand=TRUE,
                         singleEnd=FALSE, 
                         fragments=TRUE)
colSums(assay(so2))
colSums(assay(so1))
plot(assay(so1)[,2], assay(so2)[,1], xlim=c(0,5000), ylim=c(0,5000),
     xlab="single end counting", ylab="paired end counting")
abline(0,1)
abline(0,.5)
```

## Footnotes

### Methods for counting reads which overlap features

Bioconductor packages:

- `summarizeOverlaps` in the `GenomicAlignments` package

http://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html

- `featureCounts` in the `Rsubread` package

Liao Y, Smyth GK, Shi W., "featureCounts: an efficient general purpose program for assigning sequence reads to genomic features." Bioinformatics. 2014
http://www.ncbi.nlm.nih.gov/pubmed/24227677
http://bioinf.wehi.edu.au/featureCounts/

Command line tools: 

- `htseq-count`, a program in the `htseq` Python package

Simon Anders, Paul Theodor Pyl, Wolfgang Huber.
HTSeq â€” A Python framework to work with high-throughput sequencing data
bioRxiv preprint (2014), doi: [10.1101/002824](http://dx.doi.org/10.1101/002824)

http://www-huber.embl.de/users/anders/HTSeq/doc/count.html

- `bedtools` <https://code.google.com/p/bedtools/>

- `bedops` <https://code.google.com/p/bedops/>

