---
layout: page
title: Basic inference for microarray data
---

```{r options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

# Basic inference for microarray data

We have data for two strains of mice which we will refer to as strain 0 and 1. We want to know which genes are differentially expressed.  We extracted RNA from 12 randomely selected mice from each strain. In one experiment we pooled the RNA from all individuals from each strain and then created 4 replicate samples from this pool. 

```{r}
library(Biobase)
# library(devtools)
# install_github("dagdata","genomicsclass")
library(dagdata)
data(maPooling)
e <- maPooling
head(pData(e))

# install_github("rafalib","ririzarr")
library(rafalib)
mypar()
flipt <- function(m) t(m[nrow(m):1,])
myimage <- function(m,...) {
  image(flipt(m),xaxt="n",yaxt="n",...)
  }

myimage(as.matrix(pData(e)),col=c("white","black"),
        xlab="experiments",
        ylab="individuals",
        main="phenoData")

# individuals will contain the rows with data from only one experiment
individuals <- which(rowSums(pData(e)) == 1)
individuals

## remove replicates (tr = technical replicates)
names(individuals)
individuals <- individuals[-grep("tr",names(individuals))]

# graphical check: now we have only points, there are no stripes (stripe= several rows containing data from the same experiment, as in the original data).
es <- e[,individuals]
myimage(as.matrix(pData(es)),col=c("white","black"))

# strains a and b are marked with a and b in the names
es$group <- factor(as.numeric(grepl("b",colnames(es))))
es$group
```

## Plots of gene expression across group

Let's look at 2 pre-selected genes for illustration, which are the same genes from the lecture.

Expression set contains normalized gene expression values (next lectures will explain normalization, this set has already been normalized)
```{r, fig.height=3, fig.width=6}
i=11425
j=11878
mypar(1,2)
stripchart(split(exprs(es)[i,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 1", xlab="Group", pch=15)
stripchart(split(exprs(es)[j,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 2", xlab="Group", pch=15)
```

# bloxplots can also be used
boxplot(split(exprs(es)[i,], es$group))
boxplot(split(exprs(es)[j,], es$group))

# equivalent to
boxplot(exprs(es)[i,] ~ es$group)
boxplot(exprs(es)[j,] ~ es$group)

## Compute a t-test for each gene (row)

```{r}
# biocLite("genefilter")
library(genefilter)
tt <- rowttests(exprs(es), es$group)
head(tt)

# check that rowttests computes the difference between the mean of the first group and the mean of the second group, reports it tt$dm
head(tt,1)
mean(exprs(es)[1,es$group == 0]) - mean(exprs(es)[1,es$group == 1]) 

# for a single row, direct t.test calculation (rowtests will be faster for many rows)
simple.t <- t.test(exprs(es)[1,] ~ es$group, var.equal=TRUE)
simple.t$p.value
head(tt,1)

tt$p.value[i]
tt$p.value[j]

mypar(1,1)
with(tt, plot(dm, -log10(p.value), 
              xlab="difference in means",
              main="'Volcano' plot"))

# gets the value of the a point in a plot
tt[with(tt, identify(dm, -log10(p.value))),]
```

## Compare with non-parametric tests

```{r}
# take 3 samples from strain a and 3 from strain b (it is previously known that 1,2 3 belong to strain a and 13, 14, 15 belong to strain b)
es2 <- es[,c(1,2,3,13,14,15)]
head(exprs(es2))
mypar(1,1)
stripchart(exprs(es2)[1,] ~ es2$group, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(exprs(es2)[1,] ~ es2$group)
wilcox.test(exprs(es2)[1,] ~ es2$group)

# wilcox is conservative, in the following two examples we have clear separation between groups but p.value is much higher than t test
y <- 1:6
x <- es2$group
stripchart(y ~ x, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(y ~ x)
wilcox.test(y ~ x)

# very clear separation, and wilcox report conservative p-value, same as previous example, as ranks are identicals in both cases
y <- c(1:3,11:13)
stripchart(y ~ x, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(y ~ x)
wilcox.test(y ~ x)
```

## Basic inference on microarray using lmFit from limma package

We will discuss the limma package in the lecture on Hierarchical Modeling. The reference is [Smyth 2004](#foot).

```{r}
# biocLite("limma")
library(limma)
# ?lmFit
design <- model.matrix(~ es$group)
design
fit <- lmFit(es, design)
names(fit)
head(coef(fit))
# difference in means from t test is the same as the difference in means reported by t test
coef(fit)[1,]
tt[1,]
# we will introduce the eBayes() function
# in a later module called 'hierarchical modeling'
# but we call it now as it is standard in microarray analysis
fit <- eBayes(fit)
names(fit)
# p-values for intercept and difference between gene 1 and gene 2
fit$p.value[1,]
fit$t[1,]
# ebayes and t test are different
tt[1,]
# graphical comparison: in most cases they are close, but not the same
plot(-1 * tt$statistic, fit$t[,2],
     xlab="rowttests", 
     ylab="eBayes t")
abline(0,1,col="red",lwd=3)
#
head(topTable(fit, coef=2, sort.by="p"),3)
```

## Footnotes <a name="foot"></a>

Smyth GK, "Linear models and empirical bayes methods for assessing differential expression in microarray experiments". Stat Appl Genet Mol Biol. 2004 <http://www.ncbi.nlm.nih.gov/pubmed/16646809>
