---
layout: page
title: Basic inference for microarray data
---

```{r options, echo=FALSE}
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

# Basic inference for microarray data

We have data for two strains of mice which we will refer to as strain 0 and 1. We want to know which genes are differentially expressed.  We extracted RNA from 12 randomely selected mice from each strain. In one experiment we pooled the RNA from all individuals from each strain and then created 4 replicate samples from this pool. 

```{r}
library(Biobase)
# library(devtools)
# install_github("dagdata","genomicsclass")
library(dagdata)
data(maPooling)
e <- maPooling
head(pData(e))

# install_github("rafalib","ririzarr")
library(rafalib)
mypar()
flipt <- function(m) t(m[nrow(m):1,])
myimage <- function(m,...) {
  image(flipt(m),xaxt="n",yaxt="n",...)
  }

myimage(as.matrix(pData(e)),col=c("white","black"),
        xlab="experiments",
        ylab="individuals",
        main="phenoData")

individuals <- which(rowSums(pData(e)) == 1)
individuals

## remove replicates
names(individuals)
individuals <- individuals[-grep("tr",names(individuals))]

es <- e[,individuals]
myimage(as.matrix(pData(es)),col=c("white","black"))

es$group <- factor(as.numeric(grepl("b",colnames(es))))
es$group
```

## Plots of gene expression across group

Let's look at 2 pre-selected genes for illustration, which are the same genes from the lecture.

```{r, fig.height=3, fig.width=6}
i=11425
j=11878
mypar(1,2)
stripchart(split(exprs(es)[i,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 1", xlab="Group", pch=15)
stripchart(split(exprs(es)[j,], es$group), vertical=TRUE, method="jitter", col=c(1,2), main="Gene 2", xlab="Group", pch=15)
```

## Compute a t-test for each gene (row)

```{r}
# biocLite("genefilter")
library(genefilter)
tt <- rowttests(exprs(es), es$group)
head(tt)
head(tt,1)

mean(exprs(es)[1,es$group == 0]) - mean(exprs(es)[1,es$group == 1]) 

simple.t <- t.test(exprs(es)[1,] ~ es$group, var.equal=TRUE)
simple.t$p.value

tt$p.value[i]
tt$p.value[j]

mypar(1,1)
with(tt, plot(dm, -log10(p.value), 
              xlab="difference in means",
              main="'Volcano' plot"))
tt[with(tt, identify(dm, -log10(p.value))),]
```

## Compare with non-parametric tests

```{r}
es2 <- es[,c(1,2,3,13,14,15)]
mypar(1,1)
stripchart(exprs(es2)[1,] ~ es2$group, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(exprs(es2)[1,] ~ es2$group)
wilcox.test(exprs(es2)[1,] ~ es2$group)

y <- 1:6
x <- es2$group
stripchart(y ~ x, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(y ~ x)
wilcox.test(y ~ x)

y <- c(1:3,11:13)
stripchart(y ~ x, vertical=TRUE, method="jitter", col=c(1,2), main="three samples per group", xlab="Group", ylab="", pch=15)
t.test(y ~ x)
wilcox.test(y ~ x)
```

## Basic inference on microarray using lmFit from limma package

We will discuss the limma package in the lecture on Hierarchical Modeling. The reference is [Smyth 2004](#foot).

```{r}
# biocLite("limma")
library(limma)
# ?lmFit
design <- model.matrix(~ es$group)
design
fit <- lmFit(es, design)
names(fit)
head(coef(fit))
tt[1,]
# we will introduce the eBayes() function
# in a later module called 'hierarchical modeling'
# but we call it now as it is standard in microarray analysis
fit <- eBayes(fit)
names(fit)
fit$p.value[1,]
fit$t[1,]
tt[1,]
plot(-1 * tt$statistic, fit$t[,2],
     xlab="rowttests", 
     ylab="eBayes t")
abline(0,1,col="red",lwd=3)
head(topTable(fit, coef=2, sort.by="p"),3)
```

## Footnotes <a name="foot"></a>

Smyth GK, "Linear models and empirical bayes methods for assessing differential expression in microarray experiments". Stat Appl Genet Mol Biol. 2004 <http://www.ncbi.nlm.nih.gov/pubmed/16646809>
